spec/connection_spec.rb

```diff
@@ -134,6 +138,18 @@ def handle_client(client)
       expect(response).to eq "hello!"
     end
 
+    it "re-opens the connection if it is idle for more than 5 minutes" do
+      Timecop.freeze do
+        connection.send_request(request)
+
+        Timecop.travel(60 * 6)
+
```

Comment:
I knew you'd say that! ;-)

I'd rather not expose the idleness stuff at this layer, but it's difficult to test otherwise. Especially since this test actually uses the full TCP stack rather than a test double socket...

I'm thinking about splitting this class into two: Session and Connection. The former would handle lifecycle events while the latter would wrap the socket and keep track of correlation ids, last-request timestamps, etc. Connection would expose methods that Session could use to make decisions.